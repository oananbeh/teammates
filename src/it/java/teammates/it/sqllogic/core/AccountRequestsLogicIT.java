package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.UUID;

import org.testng.annotations.Test;

import teammates.common.datatransfer.AccountRequestStatus;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountRequestsLogic;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlentity.AccountRequest;

/**
 * SUT: {@link AccountRequestsLogic}.
 */
public class AccountRequestsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();

    @Test
    public void testGetAccountRequest_nonExistentAccountRequest_returnsNull() {
        UUID id = UUID.randomUUID();
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertNull(actualAccountRequest);
    }

    @Test
    public void testGetAccountRequest_existingAccountRequest_getsSuccessfully() throws InvalidParametersException {
        AccountRequest expectedAccountRequest =
                new AccountRequest("test@gmail.com", "name", "institute", AccountRequestStatus.PENDING, "comments");
        UUID id = expectedAccountRequest.getId();
        accountRequestsLogic.createAccountRequest(expectedAccountRequest);
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertEquals(expectedAccountRequest, actualAccountRequest);
    }

    @Test
    public void testResetAccountRequest()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        ______TS("success: create account request and update registeredAt field");

        String name = "name lee";
        String email = "email@gmail.com";
        String institute = "institute";
        AccountRequestStatus status = AccountRequestStatus.PENDING;
        String comments = "comments";

        AccountRequest toReset = accountRequestsLogic.createAccountRequest(name, email, institute, status, comments);
        AccountRequestsDb accountRequestsDb = AccountRequestsDb.inst();

        toReset.setRegisteredAt(Instant.now());
        UUID id = toReset.getId();
        toReset = accountRequestsDb.getAccountRequest(id);

        assertNotNull(toReset);
        assertNotNull(toReset.getRegisteredAt());

        ______TS("success: reset account request that already exists");

        AccountRequest resetted = accountRequestsLogic.resetAccountRequest(id);

        assertNull(resetted.getRegisteredAt());

        ______TS("success: test delete account request");

        accountRequestsLogic.deleteAccountRequest(toReset.getId());

        assertNull(accountRequestsLogic.getAccountRequest(toReset.getId()));

        ______TS("failure: reset account request that does not exist");

        assertThrows(EntityDoesNotExistException.class,
                () -> accountRequestsLogic.resetAccountRequest(id));
    }
}
