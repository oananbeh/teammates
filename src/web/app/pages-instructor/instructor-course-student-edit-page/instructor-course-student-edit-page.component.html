<tm-loading-retry [shouldShowRetry]="hasStudentLoadingFailed" [message]="'Failed to load student'" (retryEvent)="loadStudentEditDetails(courseId, studentEmail)">
  <div *tmIsLoading="isStudentLoading">
    <div *ngIf="student" class="card border-primary">
      <div class="card-body fill-plain text-sm-start">
        <form
          id="instructor-student-edit-form"
          class="form form-horizontal"
          (ngSubmit)="onSubmit( resendPastLinksModal)" [formGroup]="editForm">
          <div class="row mb-3">
            <label class="col-sm-2">Course ID:</label>
            <div class="col-sm-10" id="course-id">{{ courseId }}</div>
          </div>
          <div class="row mb-3">
            <label for="student-name" class="col-sm-2 col-form-label">Student Name:</label>
            <div class="col-sm-10">
              <input
                class="form-control"
                id="student-name"
                formControlName="student-name">
                <div
                  *ngIf="editForm.get('student-name')!.errors
                  && (editForm.get('student-name')!.dirty || editForm.get('student-name')!.touched)"
                  class="invalid-field">
                  <i class="fa fa-exclamation-circle" aria-hidden="true"></i>
                  <span [hidden]="!editForm.get('student-name')!.errors!['required']">
                    {{ displayEmptyFieldMessage('Student Name') }}
                  </span>
                  <span [hidden]="!editForm.get('student-name')!.errors!['maxlength']">
                    {{ displayExceedMaxLengthMessage('Student Name', FormValidator.STUDENT_NAME_MAX_LENGTH) }}
                  </span>
                </div>
            </div>
          </div>
          <div class="row mb-3">
            <label for="section-name" class="col-sm-2 col-form-label">Section Name:</label>
            <div class="col-sm-10">
              <input
                class="form-control"
                id="section-name"
                formControlName="section-name">
              <div
                *ngIf="editForm.get('section-name')!.errors
                && (editForm.get('section-name')!.dirty || editForm.get('section-name')!.touched)"
                class="invalid-field">
                <i class="fa fa-exclamation-circle" aria-hidden="true"></i>
                <span [hidden]="!editForm.get('section-name')!.errors!['required']">
                  {{ displayEmptyFieldMessage('Section Name') }}
                </span>
                <span [hidden]="!editForm.get('section-name')!.errors!['maxlength']">
                  {{ displayExceedMaxLengthMessage('Section Name', FormValidator.SECTION_NAME_MAX_LENGTH) }}
                </span>
              </div>
            </div>
          </div>
          <div class="row mb-3">
            <label for="team-name" class="col-sm-2 col-form-label">Team Name:</label>
            <div class="col-sm-10">
              <input
                class="form-control col-form-label"
                name="team-name"
                id="team-name"
                formControlName="team-name">
              <div
                *ngIf="editForm.get('team-name')!.errors
                && (editForm.get('team-name')!.dirty || editForm.get('team-name')!.touched)"
                class="invalid-field">
                <i class="fa fa-exclamation-circle" aria-hidden="true"></i>
                <span [hidden]="!editForm.get('team-name')!.errors!['required']">
                  {{ displayEmptyFieldMessage('Team Name') }}
                </span>
                <span [hidden]="!editForm.get('team-name')!.errors!['maxlength']">
                  {{ displayExceedMaxLengthMessage('Team Name', FormValidator.TEAM_NAME_MAX_LENGTH) }}
                </span>
              </div>
            </div>
          </div>
          <div class="row mb-3">
            <label for="new-student-email" class="col-sm-2 col-form-label">E-mail Address:</label>
            <div class="col-sm-10">
              <input
                class="form-control"
                id="new-student-email"
                formControlName="new-student-email">
              <div
                *ngIf="editForm.get('new-student-email')!.errors
                && (editForm.get('new-student-email')!.dirty || editForm.get('new-student-email')!.touched)"
                  class="invalid-field">
                <i class="fa fa-exclamation-circle" aria-hidden="true"></i>
                <span [hidden]="!editForm.get('new-student-email')!.errors!['required']">
                  {{ displayEmptyFieldMessage('E-mail Address') }}
                </span>
                <span [hidden]="!editForm.get('new-student-email')!.errors!['maxlength']">
                  {{ displayExceedMaxLengthMessage('E-mail Address', FormValidator.EMAIL_MAX_LENGTH) }}
                </span>
              </div>
            </div>
          </div>
          <div class="row mb-3">
            <label for="comments" class="col-sm-2 col-form-label">Comments:</label>
            <div class="col-sm-10">
              <textarea
                class="form-control"
                rows="6"
                id="comments"
                formControlName="comments">
              </textarea>
            </div>
          </div>
          <br>
          <div class="text-center">
            <button class="btn btn-success"
                    id="btn-submit"
                    type="submit"
                    [disabled]="!editForm.valid || isFormSaving">
              <i [ngClass]="isFormSaving ? 'spinner-border spinner-border-sm' : 'fas fa-save'"></i>
              Save Changes
            </button>
          </div>
          <br>
          <br>

          <ng-template #resendPastLinksModal let-modal>
            <div class="modal-header alert alert-primary alert-no-bottom">
              <h2 class="h4 modal-title icon-primary">Resend past links to the new email?</h2>
              <button
                  type="button"
                  class="btn-close"
                  data-dismiss="modal"
                  aria-hidden="true"
                  (click)="modal.dismiss()"></button>
            </div>
            <div class="modal-body">
              Do you want to resend past session links of this course to the new email {{ editForm.get('newstudentemail')?.value }}?
            </div>
            <div class="modal-footer">
              <button id="btn-resend-links" type="button" class="btn btn-primary"
                      (click)="submitEditForm(true); modal.dismiss()">
                Yes, save changes and resend links
              </button>
              <button type="button" class="btn btn-primary"
                      (click)="submitEditForm(false); modal.dismiss()">
                No, just save the changes
              </button>
              <button type="button" class="btn btn-outline-secondary"
                      (click)="modal.dismiss()">
                Cancel
              </button>
            </div>
          </ng-template>
        </form>
      </div>
    </div>
  </div>
</tm-loading-retry>
