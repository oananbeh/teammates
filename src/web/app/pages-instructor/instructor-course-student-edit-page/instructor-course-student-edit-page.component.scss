.fill-plain {
  background-color: #eaeff5;
}

label {
  font-weight: bold;
}

.invalid-field {
  padding-top: 5px;
  color: #b50000;
}
