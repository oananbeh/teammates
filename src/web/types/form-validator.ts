import { ApiConst, ApiStringConst } from './api-const';

/**
 * Represents the root FormValidator object of all form fields.
 */
export enum FormValidator {
  /**
   * Max length for the 'Course ID' field.
   */
  COURSE_ID_MAX_LENGTH = ApiConst.COURSE_ID_MAX_LENGTH,

  /**
   * Max length for the 'Course Name' field.
   */
  COURSE_NAME_MAX_LENGTH = ApiConst.COURSE_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Student Name` field.
   */
  STUDENT_NAME_MAX_LENGTH = ApiConst.STUDENT_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Section Name` field.
   */
  SECTION_NAME_MAX_LENGTH = ApiConst.SECTION_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Team Name` field.
   */
  TEAM_NAME_MAX_LENGTH = ApiConst.TEAM_NAME_MAX_LENGTH,

  /**
   * Max length for the 'E-mail Address` field.
   */
  EMAIL_MAX_LENGTH = ApiConst.EMAIL_MAX_LENGTH,

  /**
   * Regex used to verify emails in the back-end.
   */
  EMAIL_REGEX = ApiStringConst.EMAIL_REGEX,

  /**
   * Regex used to verify names.
   *
   * Based on back-end's `FieldValidator.REGEX_NAME`.
   * The back-end regex is not converted to use here as the pattern syntax is not accepted in JS.
   */
  NAME_REGEX = '^[a-zA-Z0-9][^|%]*$',

  /**
   * Regex used to verify country names.
   *
   * Based on back-end's `FieldValidator.REGEX_NAME`, but without needing to start with alphanumeric
   * as the country is added to the end of the combined institute string.
   */
  COUNTRY_REGEX = '^[^|%]*$',

  /**
   * Max length for institution name in account request. (to be combined with country)
   */
  INSTITUTION_NAME_MAX_LENGTH = 86,

  /**
   * Max length for country in account request. (to be combined with institution name)
   */
  COUNTRY_NAME_MAX_LENGTH = 40,
}
