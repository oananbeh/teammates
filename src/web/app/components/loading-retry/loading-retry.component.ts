import { Component, EventEmitter, Input, Output } from '@angular/core';

/**
 * Retry button for when loading fails.
 */
@Component({
  selector: 'tm-loading-retry',
  templateUrl: './loading-retry.component.html',
  styleUrls: ['./loading-retry.component.scss'],
})
export class LoadingRetryComponent {

  @Input() message: string = '';

  @Input() shouldShowRetry: boolean = false;

  @Input() retryButtonDisabled: boolean = false;

  @Output() retryEvent: EventEmitter<any> = new EventEmitter<any>();

}
