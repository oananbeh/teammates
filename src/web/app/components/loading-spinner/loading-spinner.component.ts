import { Component } from '@angular/core';

/**
 * Loading spinner to show when waiting for request.
 */
@Component({
  selector: 'tm-loading-spinner',
  templateUrl: './loading-spinner.component.html',
  styleUrls: ['./loading-spinner.component.scss'],
})
export class LoadingSpinnerComponent {
  loadingMessage: string = 'Loading...';

  constructor() {
    // After 5 seconds of loading show message to assure user the page is not hanging.
    setTimeout(() => { this.loadingMessage = 'Still loading...'; }, 5000);
  }
}
