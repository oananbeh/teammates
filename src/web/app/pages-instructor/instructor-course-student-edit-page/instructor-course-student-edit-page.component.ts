import { Component, Input, OnDestroy, OnInit } from '@angular/core';
import { AbstractControl, UntypedFormControl, UntypedFormGroup, Validators } from '@angular/forms';
import { ActivatedRoute } from '@angular/router';
import { NgbModal, NgbModalRef } from '@ng-bootstrap/ng-bootstrap';
import { Subscription } from 'rxjs';
import { finalize } from 'rxjs/operators';

import { NavigationService } from '../../../services/navigation.service';
import { SimpleModalService } from '../../../services/simple-modal.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { StudentService } from '../../../services/student.service';
import { JoinState, MessageOutput, Student } from '../../../types/api-output';
import { StudentUpdateRequest } from '../../../types/api-request';
import { FormValidator } from '../../../types/form-validator';
import { SimpleModalType } from '../../components/simple-modal/simple-modal-type';
import { ErrorMessageOutput } from '../../error-message-output';

/**
 * Instructor course student edit page.
 */
@Component({
  selector: 'tm-instructor-course-student-edit-page',
  templateUrl: './instructor-course-student-edit-page.component.html',
  styleUrls: ['./instructor-course-student-edit-page.component.scss'],
})
export class InstructorCourseStudentEditPageComponent implements OnInit, OnDestroy {

  FormValidator: typeof FormValidator = FormValidator; // enum

  @Input() isEnabled: boolean = true;
  courseId: string = '';
  studentEmail: string = '';
  student!: Student;

  isTeamnameFieldChanged: boolean = false;
  isEmailFieldChanged: boolean = false;
  isStudentLoading: boolean = false;
  hasStudentLoadingFailed: boolean = false;
  isFormSaving: boolean = false;

  editForm!: UntypedFormGroup;
  teamFieldSubscription?: Subscription;
  emailFieldSubscription?: Subscription;

  constructor(private route: ActivatedRoute,
              private statusMessageService: StatusMessageService,
              private studentService: StudentService,
              private navigationService: NavigationService,
              private ngbModal: NgbModal,
              private simpleModalService: SimpleModalService) { }

  ngOnInit(): void {
    if (!this.isEnabled) {
      this.student = {
        email: 'alice@email.com',
        courseId: '',
        name: 'Alice Betsy',
        comments: 'Alice is a transfer student.',
        teamName: 'Team A',
        sectionName: 'Section A',
        joinState: JoinState.JOINED,
      };
      this.initEditForm();
      return;
    }

    this.route.queryParams.subscribe((queryParams: any) => {
      this.courseId = queryParams.courseid;
      this.studentEmail = queryParams.studentemail;
      this.loadStudentEditDetails(queryParams.courseid, queryParams.studentemail);
    });
  }

  ngOnDestroy(): void {
    if (this.emailFieldSubscription) {
      (this.emailFieldSubscription as Subscription).unsubscribe();
    }
    if (this.teamFieldSubscription) {
      (this.teamFieldSubscription as Subscription).unsubscribe();
    }
  }

  /**
   * Loads student details required for this page.
   */
  loadStudentEditDetails(courseId: string, studentEmail: string): void {
    this.hasStudentLoadingFailed = false;
    this.isStudentLoading = true;
    this.studentService.getStudent(
        courseId, studentEmail,
    ).pipe(finalize(() => {
      this.isStudentLoading = false;
    })).subscribe({
      next: (student: Student) => {
        this.student = student;
        this.initEditForm();
      },
      error: (resp: ErrorMessageOutput) => {
        this.hasStudentLoadingFailed = true;
        this.statusMessageService.showErrorToast(resp.error.message);
      },
    });
  }

  /**
   * Initializes the student details edit form with the fields fetched from the backend.
   * Subscriptions are set up to listen to changes in the 'teamname' fields and 'newstudentemail' fields.
   */
  private initEditForm(): void {
    this.editForm = new UntypedFormGroup({
      'student-name': new UntypedFormControl(this.student.name,
          [Validators.required, Validators.maxLength(FormValidator.STUDENT_NAME_MAX_LENGTH)]),
      'section-name': new UntypedFormControl(this.student.sectionName,
          [Validators.required, Validators.maxLength(FormValidator.SECTION_NAME_MAX_LENGTH)]),
      'team-name': new UntypedFormControl(this.student.teamName,
          [Validators.required, Validators.maxLength(FormValidator.TEAM_NAME_MAX_LENGTH)]),
      'new-student-email': new UntypedFormControl(this.student.email, // original student email initialized
          [Validators.required, Validators.maxLength(FormValidator.EMAIL_MAX_LENGTH)]),
      comments: new UntypedFormControl(this.student.comments),
    });
    this.teamFieldSubscription =
        (this.editForm.get('team-name') as AbstractControl).valueChanges
            .subscribe(() => {
              this.isTeamnameFieldChanged = true;
            });

    this.emailFieldSubscription =
        (this.editForm.get('new-student-email') as AbstractControl).valueChanges
            .subscribe(() => {
              this.isEmailFieldChanged = true;
            });
  }

  /**
   * Displays message to user stating that the field is empty.
   */
  displayEmptyFieldMessage(fieldName: string): string {
    return `The field '${fieldName}' should not be empty.`;
  }

  /**
   * Displays message to user stating that the field exceeds the max length.
   */
  displayExceedMaxLengthMessage(fieldName: string, maxLength: number): string {
    return `The field '${fieldName}' should not exceed ${maxLength} characters.`;
  }

  /**
   * Handles logic related to showing the appropriate modal boxes
   * upon submission of the form. Submits the form otherwise.
   */
  onSubmit(resendPastLinksModal: any): void {
    if (!this.isEnabled) {
      return;
    }

    if (this.isTeamnameFieldChanged) {
      const modalContent: string =
          `Editing these fields will result in some existing responses from this student to be deleted.
          You may download the data before you make the changes.`;
      const modalRef: NgbModalRef = this.simpleModalService.openConfirmationModal(
          'Delete existing responses?', SimpleModalType.WARNING, modalContent);
      modalRef.result.then(() => {
        this.deleteExistingResponses(resendPastLinksModal);
      }, () => {});
    } else if (this.isEmailFieldChanged) {
      this.ngbModal.open(resendPastLinksModal);
    } else {
      this.submitEditForm(false);
    }
  }

  /**
   * Shows the `resendPastSessionLinks` modal if email field has changed.
   * Submits the form  otherwise.
   */
  deleteExistingResponses(resendPastLinksModal: any): void {
    if (this.isEmailFieldChanged) {
      this.ngbModal.open(resendPastLinksModal);
    } else {
      this.submitEditForm(false);
    }
  }

  /**
   * Submits the form data to edit the student details.
   */
  submitEditForm(shouldResendPastSessionLinks: boolean): void {
    const reqBody: StudentUpdateRequest = {
      name: this.editForm.value['student-name'],
      email: this.editForm.value['new-student-email'],
      team: this.editForm.value['team-name'],
      section: this.editForm.value['section-name'],
      comments: this.editForm.value.comments,
      isSessionSummarySendEmail: shouldResendPastSessionLinks,
    };

    this.isFormSaving = true;

    this.studentService.updateStudent({
      courseId: this.courseId,
      studentEmail: this.student.email,
      requestBody: reqBody,
    })
      .pipe(finalize(() => {
        this.isFormSaving = false;
      }))
      .subscribe({
        next: (resp: MessageOutput) => {
          this.navigationService.navigateWithSuccessMessage('/web/instructor/courses/details',
              resp.message, { courseid: this.courseId });
        },
        error: (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      });
  }
}
