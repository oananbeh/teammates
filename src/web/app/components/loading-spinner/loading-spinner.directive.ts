import {
  ComponentRef,
  Directive,
  Input,
  TemplateRef,
  ViewContainerRef,
} from '@angular/core';
import { LoadingSpinnerComponent } from './loading-spinner.component';

/**
 * Directive for loading spinner component
 */
@Directive({
  selector: '[tmIsLoading]',
})
export class LoadingSpinnerDirective {
  loadingComponent: ComponentRef<LoadingSpinnerComponent>;

  constructor(private templateRef: TemplateRef<any>,
              private viewContainer: ViewContainerRef) {
    this.loadingComponent = this.viewContainer.createComponent(LoadingSpinnerComponent);
  }

  @Input() set tmIsLoading(loading: boolean) {
    this.viewContainer.clear();

    if (loading) {
      this.loadingComponent = this.viewContainer.createComponent(LoadingSpinnerComponent);
    } else {
      this.viewContainer.createEmbeddedView(this.templateRef);
    }
  }
}
