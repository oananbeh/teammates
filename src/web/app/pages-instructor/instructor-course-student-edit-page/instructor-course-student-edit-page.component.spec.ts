import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { ReactiveFormsModule, UntypedFormControl, UntypedFormGroup } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { InstructorCourseStudentEditPageComponent } from './instructor-course-student-edit-page.component';
import { JoinState } from '../../../types/api-output';
import { LoadingRetryModule } from '../../components/loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';

describe('InstructorCourseStudentEditPageComponent', () => {
  let component: InstructorCourseStudentEditPageComponent;
  let fixture: ComponentFixture<InstructorCourseStudentEditPageComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [InstructorCourseStudentEditPageComponent],
      imports: [
        RouterTestingModule,
        ReactiveFormsModule,
        HttpClientTestingModule,
        LoadingSpinnerModule,
        LoadingRetryModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorCourseStudentEditPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with student details', () => {
    component.student = {
      email: 'jake@gmail.com',
      courseId: 'Crime101',
      name: 'Jake Peralta',
      comments: 'Cool cool cool.',
      teamName: 'Team A',
      sectionName: 'Section A',
      joinState: JoinState.JOINED,
    };
    component.editForm = new UntypedFormGroup({
      'student-name': new UntypedFormControl('Jake Peralta'),
      'section-name': new UntypedFormControl('Section A'),
      'team-name': new UntypedFormControl('Team A'),
      'new-student-email': new UntypedFormControl('jake@gmail.com'),
      comments: new UntypedFormControl('Cool cool cool.'),
    });
    component.isStudentLoading = false;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when student is still loading', () => {
    component.isStudentLoading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
