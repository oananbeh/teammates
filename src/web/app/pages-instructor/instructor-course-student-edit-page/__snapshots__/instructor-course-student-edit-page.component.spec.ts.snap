// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorCourseStudentEditPageComponent should snap when student is still loading 1`] = `
<tm-instructor-course-student-edit-page
  FormValidator={[Function Object]}
  courseId="undefined"
  hasStudentLoadingFailed="false"
  isEmailFieldChanged="false"
  isEnabled={[Function Boolean]}
  isFormSaving="false"
  isStudentLoading={[Function Boolean]}
  isTeamnameFieldChanged="false"
  navigationService={[Function NavigationService]}
  ngbModal={[Function _NgbModal]}
  route={[Function ActivatedRoute]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentEmail="undefined"
  studentService={[Function StudentService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry>
</tm-instructor-course-student-edit-page>
`;

exports[`InstructorCourseStudentEditPageComponent should snap with default fields 1`] = `
<tm-instructor-course-student-edit-page
  FormValidator={[Function Object]}
  courseId="undefined"
  hasStudentLoadingFailed="false"
  isEmailFieldChanged="false"
  isEnabled={[Function Boolean]}
  isFormSaving="false"
  isStudentLoading={[Function Boolean]}
  isTeamnameFieldChanged="false"
  navigationService={[Function NavigationService]}
  ngbModal={[Function _NgbModal]}
  route={[Function ActivatedRoute]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentEmail="undefined"
  studentService={[Function StudentService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry>
</tm-instructor-course-student-edit-page>
`;

exports[`InstructorCourseStudentEditPageComponent should snap with student details 1`] = `
<tm-instructor-course-student-edit-page
  FormValidator={[Function Object]}
  courseId="undefined"
  editForm={[Function FormGroup]}
  hasStudentLoadingFailed="false"
  isEmailFieldChanged="false"
  isEnabled={[Function Boolean]}
  isFormSaving="false"
  isStudentLoading="false"
  isTeamnameFieldChanged="false"
  navigationService={[Function NavigationService]}
  ngbModal={[Function _NgbModal]}
  route={[Function ActivatedRoute]}
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  student={[Function Object]}
  studentEmail="undefined"
  studentService={[Function StudentService]}
>
  <tm-loading-retry>
    <div>
      <div
        class="card border-primary"
      >
        <div
          class="card-body fill-plain text-sm-start"
        >
          <form
            class="form form-horizontal ng-untouched ng-pristine ng-valid"
            id="instructor-student-edit-form"
            novalidate=""
          >
            <div
              class="row mb-3"
            >
              <label
                class="col-sm-2"
              >
                Course ID:
              </label>
              <div
                class="col-sm-10"
                id="course-id"
              >
              </div>
            </div>
            <div
              class="row mb-3"
            >
              <label
                class="col-sm-2 col-form-label"
                for="student-name"
              >
                Student Name:
              </label>
              <div
                class="col-sm-10"
              >
                <input
                  class="form-control ng-untouched ng-pristine ng-valid"
                  formcontrolname="student-name"
                  id="student-name"
                />
              </div>
            </div>
            <div
              class="row mb-3"
            >
              <label
                class="col-sm-2 col-form-label"
                for="section-name"
              >
                Section Name:
              </label>
              <div
                class="col-sm-10"
              >
                <input
                  class="form-control ng-untouched ng-pristine ng-valid"
                  formcontrolname="section-name"
                  id="section-name"
                />
              </div>
            </div>
            <div
              class="row mb-3"
            >
              <label
                class="col-sm-2 col-form-label"
                for="team-name"
              >
                Team Name:
              </label>
              <div
                class="col-sm-10"
              >
                <input
                  class="form-control col-form-label ng-untouched ng-pristine ng-valid"
                  formcontrolname="team-name"
                  id="team-name"
                  name="team-name"
                />
              </div>
            </div>
            <div
              class="row mb-3"
            >
              <label
                class="col-sm-2 col-form-label"
                for="new-student-email"
              >
                E-mail Address:
              </label>
              <div
                class="col-sm-10"
              >
                <input
                  class="form-control ng-untouched ng-pristine ng-valid"
                  formcontrolname="new-student-email"
                  id="new-student-email"
                />
              </div>
            </div>
            <div
              class="row mb-3"
            >
              <label
                class="col-sm-2 col-form-label"
                for="comments"
              >
                Comments:
              </label>
              <div
                class="col-sm-10"
              >
                <textarea
                  class="form-control ng-untouched ng-pristine ng-valid"
                  formcontrolname="comments"
                  id="comments"
                  rows="6"
                >
                </textarea>
              </div>
            </div>
            <br />
            <div
              class="text-center"
            >
              <button
                class="btn btn-success"
                id="btn-submit"
                type="submit"
              >
                <i
                  class="fas fa-save"
                />
                 Save Changes 
              </button>
            </div>
            <br />
            <br />
          </form>
        </div>
      </div>
    </div>
  </tm-loading-retry>
</tm-instructor-course-student-edit-page>
`;
